src/riscv/linux/init.c

```diff
@@ -0,0 +1,112 @@
+#include <string.h>
+
+#include <cpuinfo/internal-api.h>
+#include <cpuinfo/log.h>
+#include <linux/api.h>
+#include <riscv/linux/api.h>
+
+#define ALLOCATE_OR_FAIL(dest, count, size, description) \
```

Comment:
No macros please, just duplicate the code. This code will be read way more often than it is written, and avoid indirections helps understanding.